node: Replace ipv[46]MasqAddrs with Table[NodeAddress]

joamaki opened the pull request:
Replace the global map of BPF masquerade addresses with a query to Table[NodeAddress] in the loader's patchHostNetdevDatapath and remove the now unused code around masquerade addresses in pkg/node.

joamaki committed:
node: Replace ipv[46]MasqAddrs with Table[NodeAddress]

Replace the global map of BPF masquerade addresses with a query to
Table[NodeAddress] in the loader's patchHostNetdevDatapath and
remove the now unused code around masquerade addresses in pkg/node.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki commented on cilium-dbg/cmd/cleanup.go:
```diff
@@ -573,8 +573,13 @@ func removeTCFilters(linkAndFilters map[string][]*netlink.BpfFilter) error {
 }
 
 func removeXDPAttachments(links []netlink.Link) error {
+	params := loader.Params{
```
Very ugly, but we've got follow-ups coming right after that clean this up by moving the DetachXDP into a  function.

ti-mo replied:
The only thing I would change is shortening to `loader.NewLoader(loader.Params{...})`.. :p Not sure what's so bad about this, we need to pull config from somewhere and we need to instantiate loader somehow. We could always make the argument to `NewLoader()` nillable.

joamaki replied:
This will become impossible to instantiate in `cilium-dbg` pretty soon, but it's fine for now.

joamaki commented on pkg/datapath/loader/cell.go:
```diff
@@ -38,6 +41,25 @@ type Loader interface {
+	// DeriveMasqIPAddrFromDevice specifies which devices IP addr is used for BPF masquerade.
+	// See commit d204d789746b1389cc2ba02fdd55b81a2f55b76e for original context. This can
+	// be removed once https://github.com/cilium/cilium/issues/17158 is resolved.
+	DeriveMasqIPAddrFromDevice string
```
I feel somewhat weird about introducing this Config for the loader, but I wanted to have this config option as close as possible to its use-site. Open to suggestions to do this in some other way.

ti-mo replied:
All good. Ack on the comment adding context, that's very useful!

A config flag starting with a verb (`Derive` in this case) typically conveys that this is a boolean, though. (e.g. derive the masq IP or not?) Also e2manywords; let's call this what it is: `MasqueradeDevice`. The fact that it's (currently) used to derive the masq IP can go in a comment on the field, though if it's extended later, that will go stale.

Unless we can tackle #17158 in the very short term (1.16) and it's actually planned, let's not assume that this is temporary and just document this as a feature/requirement of the loader.

One more thing, this string can be empty. It's not clear what the caller should expect when it's left empty. If it's optional, it needs to be specified why.

dylandreimerink replied:
Adding the config here feels strange indeed. But I think we can lift it out of the loader once we split the device logic, so it hopefully will not be here for long.

joamaki commented on pkg/datapath/loader/loader.go:
```diff
@@ -135,15 +156,34 @@ func nullifyStringSubstitutions(strings map[string]string) map[string]string {
 }
 
+func (l *loader) bpfMasqAddrs(ifName string) (ipv4, ipv6 netip.Addr) {
+	if l.cfg.DeriveMasqIPAddrFromDevice != "" {
```
We're doing a one-off query here for the addresses. These of course can change later which would require reconciliation (which is missing from this PR). This will be done in a later PR that @dylandreimerink is working on. (We'll likely pull the query to the "top-half" of the loader and have a reconciliation loop there). Here as well I'm open to doing that refactoring already by instead passing the BPF masquerade addresses in some state struct down to `patchHostNetdevDatapath`. For now I opted not to do that yet as the design isn't entirely baked yet for it.

ti-mo replied:
Adding reconciliation later is perfectly fine, of course. Though I'd prefer if the query happens as early as possible so we can get rid of the stringly input param asap and pass around link objects instead. Not sure if there's some kind of startup ordering here, I assume this device exists before the agent starts?

ti-mo commented on pkg/datapath/linux/config/config_test.go:
```diff
@@ -48,6 +50,22 @@ var (
+func testParams() loader.Params {
+	nodeAddrs, err := tables.NewNodeAddressTable()
+	if err != nil {
+		panic(err)
```
`tb.Helper()`, `tb.Fatal()`, etc. etc.

ti-mo commented on pkg/datapath/loader/loader.go:
```diff
@@ -135,15 +156,34 @@ func nullifyStringSubstitutions(strings map[string]string) map[string]string {
+		tables.NodeAddressDeviceNameIndex.Query(ifName),
+	)
+	for addr, _, ok := iter.Next(); ok; addr, _, ok = iter.Next() {
+		if addr.Primary {
```
`if !addr.Primary { continue; ... }`? Whole loop body wrapped in an if statement is a bit funky. :slightly_smiling_face:

ti-mo commented on pkg/datapath/loader/loader.go:
```diff
@@ -180,14 +220,15 @@ func (l *loader) patchHostNetdevDatapath(ep datapath.Endpoint, objPath, dstPath,
-		if option.Config.EnableIPv4Masquerade && bpfMasqIPv4Addrs != nil {
-			ipv4 := bpfMasqIPv4Addrs[ifName]
-			opts["IPV4_MASQUERADE"] = uint64(byteorder.NetIPv4ToHost32(ipv4))
+		ipv4, ipv6 := l.bpfMasqAddrs(ifName)
```
```suggestion
		masq4, masq6 := l.bpfMasqAddrs(ifName)
```

ti-mo commented on pkg/datapath/linux/config/config_test.go:
```diff
@@ -48,6 +50,22 @@ var (
 	ipv6DummyAddr = netip.MustParseAddr("2001:db08:0bad:cafe:600d:bee2:0bad:cafe")
 )
 
+func testParams() loader.Params {
```
Could we do a single `TestLoader` helper in `loader/loader_test.go` that assembles the args and calls `newLoader()`?

joamaki replied:
I'll think about it. We'll likely change this many times over in follow-up PRs that add the loader reconciliation loop, but I'll try to make this as clean as I can already.

joamaki replied:
Added `NewLoaderForTest` to `loader/loader.go` and using it from both `loader_test.go` and `config_test.go` (it can't go into `loader_test.go` as then it's not accessible from `config_test.go`).

ti-mo commented on pkg/datapath/loader/loader_test.go:
```diff
@@ -42,6 +44,22 @@ var (
 	bpfDir = filepath.Join("..", "..", "..", "bpf")
 )
 
+func testParams() Params {
```
See my previous comment about `TestLoader`.

joamaki replied:
Addressed.

ti-mo reviewed (changes requested):
Thanks! Left a few minor comments.

brb commented on pkg/datapath/loader/loader.go:
```diff
@@ -135,15 +170,35 @@ func nullifyStringSubstitutions(strings map[string]string) map[string]string {
+		l.db.ReadTxn(),
+		tables.NodeAddressDeviceNameIndex.Query(ifName),
+	)
+	for addr, _, ok := iter.Next(); ok; addr, _, ok = iter.Next() {
```
Dummy q, but where does the logic of `firstGlobal()` live now?

joamaki replied:
It's still in `pkg/node/address_linux.go` as it's used by (and only by) `node.SetDefaultPrefix`. It'll be removed completely once that is refactored. 

I've been quite careful in replicating the scope-based selection logic of `firstGlobalAddr` in `pkg/datapath/tables/node_address.go`, but I would welcome another set of eyes to compare the two. 

EDIT: now that I compare again, it might be I'll need to also include the `IsPublicAddr` logic for picking `Primary`

brb replied:
> now that I compare again, it might be I'll need to also include the IsPublicAddr logic for picking Primary

That's why I was asking the question :-)

joamaki replied:
Addressed in https://github.com/cilium/cilium/pull/30579/.

asauber commented on pkg/datapath/loader/loader.go:
```diff
@@ -135,15 +170,35 @@ func nullifyStringSubstitutions(strings map[string]string) map[string]string {
+			continue
+		}
+		if addr.Addr.Is4() && masq4.IsUnspecified() {
+			masq4 = addr.Addr
```
This implicitly causes this function to return the last address from the StateDB query. Is this the intended behavior? If so there should at least be a comment here explaining why that's the case.

brb replied:
Doesn't it return the first addr? After first hit, the `masq4` is set, and then the setting is skipped for subsequent iterations.

Anyway, I think you might have scratched a subtle change in the addr selection. Previously, if the `firstGlobalAddr()` didn't find an IP addr for the given iface, then some a public IP addr of some other device was returned:
```
	        // Just make sure that we always return the same one and not a
		// random one. More info in the issue GH-7637.
		sort.SliceStable(ipsPublic, func(i, j int) bool {
			return ipsPublic[i].LinkIndex < ipsPublic[j].LinkIndex
		})
```

I wonder whether that behavior is still needed.

joamaki replied:
Added a fallback to pick some public IP :). PTAL and review the new code carefully.

brb replied:
Thanks! That code is needed for BPF NodePort too. Let's move it to `pkg/node/address.go`?

joamaki replied:
Reworked it so that now we have a "wildcard device" (`*`) in the NodeAddress table which looks at all devices to find a suitable, preferably public IP.

joamaki committed:
NodeAddress: Add wildcard device for fallback addresses

Some setups that use e.g. ECMP may not have addresses assigned to all
devices. For BPF masquerading one can use DeriveMasqIPFromDevice to work
around this, but we don't have a similar mechanism for the direct routing
address.

To fix this and retain the semantics of v1.14, add a "wildcard device"
(*) to the NodeAddress table that picks suitable fallback addresses from
all devices in the system. The criteria (in order) is to prefer public over
private, lower scope, lower ifindex and finally lower address.

To allow overlapping addresses in the NodeAddress table, use a primary key
that combines address with device name.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

asauber commented on pkg/datapath/loader/loader.go:
```diff
@@ -140,15 +176,57 @@ func nullifyStringSubstitutions(strings map[string]string) map[string]string {
+
+	find := func(iter statedb.Iterator[tables.NodeAddress]) bool {
+		for addr, _, ok := iter.Next(); ok; addr, _, ok = iter.Next() {
+			if ifName != defaults.HostDevice && addr.DeviceName == defaults.HostDevice {
```
Why not the following, instead?

```
if ifName != addr.DeviceName {
	continue
}
```

Do we suspect that if not `HostDevice`, `addr.DeviceName` and the configured `ifname` are unlikely to match?

joamaki replied:
Hmm I have no idea anymore why I added this there. The whole check isn't useful at all. Removed.

asauber reviewed (approved):
The new logic looks good to me, besides the above comment.

I'll assume that the function is moved if needed as discussed here https://github.com/cilium/cilium/pull/30457#discussion_r1495485276

brb reviewed (approved):
Thanks!

ldelossa reviewed (approved):
LGTM
